### DISCUSSION TITLE
Design Web UI Backend

### DISCUSSION THREAD

**Discussion by fkromer** [strictdoc-project/strictdoc#566] [Ideas]
This discussion thread shall clarify the software design for the feature https://github.com/strictdoc-project/strictdoc/issues/563#issuecomment-1003560436 .

As most developers interested in requirement management will have a strong background in Python (usually as their main second programming language) I would use Python for the backend. As strictdoc is implemented in Python this eases deployment of the backend in the development context (running the backend framework in development mode). In addition the backend could be extended for deployment as standalone docker container.

As backend framework I‘d suggest FastAPI. It is lightweight as Flask but profits from pydantic w.r.t. code reliability. Using Django would be overkill.

For the frontend integration part I‘d suggest GraphQL. It allows to define a typed interface and leads to less boilerplate in comparison to REST. A framework resulting in little code is Strawberry. In case it would be limited w.r.t. features one would have to migrate to Ariadne or Graphene.

W.r.t. the data source integration part it‘s probably best to consider reqif files as single source of truth and read/write from it using [reqif](https://github.com/strictdoc-project/reqif). This implies the need for converting between reqif file and sdoc file before reading and after editing data with the backend. However reqif seems to be the only standardized requirement interchange format. In addition this allows a complete decoupling of strictdoc core functionality and the web UI backend.

**Comment by fkromer** [comment]
@stanislaw I could find the class representations of ReqIf "entities" in [`reqif/reqif/models`](https://github.com/strictdoc-project/reqif/tree/main/reqif/models). For reading a ReqIf file I should be fine with [`ReqIFParser.parse(input_path: str)`](https://github.com/strictdoc-project/reqif/blob/e95bd575dc7e76a97f61a548acf1a0223add19bc/reqif/parser.py#L46-L61), right? The returned [`ReqIFBundle`](https://github.com/strictdoc-project/reqif/blob/e95bd575dc7e76a97f61a548acf1a0223add19bc/reqif/reqif_bundle.py#L16) should contain all the "entities"...

**Comment by fkromer** [comment]
After doing some research about ReqIF viewers I found [reqifviewer](https://github.com/RHEAGROUP/reqifviewer) which can visualize `reqif` file content. It's a C# / Blazor App using [reqifsharp](https://github.com/RHEAGROUP/reqifsharp) under the hood. There is a hosted version of the viewer https://viewer.reqifsharp.org/ and it can be run locally using docker. As long as `reqif` files generated from `SDoc` files can be visualized and the viewer provides all functionality required there is no need for reinventing the wheel.

**Reply by gioboske** [reply]
Do you plan to integrate this ReqIF web viewer in `SDoc`?
I've seen that a ["Web server and editable HTML pages"](https://strictdoc.readthedocs.io/en/stable/strictdoc.html#advanced) is somehow considered in the `SDoc` roadmap. This would be a killer functionality!

I'd like to know if this discussion can be considered the first step to reach that milestone.

Thanks!

**Reply by stanislaw** [reply]
This is a very good question and I also consider the Web server and UI for editing the requirements to be one of the most wanted Killer features.

While the reqifviewer might be a solution that works well for @fkromer, I don't think that this tool satisfies the requirement of "Web server and editable HTML pages" for StrictDoc.

With a limited time that I have for working on this project, the UI work has been postponed several times. I don't know if I could start working on this soon, but if I did I would likely use a software stack very similar to the one suggested by @fkromer:

FastAPI, GraphQL and probably Vue.js for the frontend part.

I would be happy to share this effort with anyone who has frontend development experience.

**Reply by gioboske** [reply]
@stanislaw thank you for the clarification!
Unfortunately, I'm not skilled in frontend developing, even if start dealing with FastAPi and Vue.js is in my to-do list since long time...

Let me broaden the discussion (a little)...
Do you think it is needed to reconsider the current static-generation approach of `SDoc` (as mentioned in the [Backlog](https://strictdoc.readthedocs.io/en/stable/strictdoc.html#advanced) comments)?

Also (even if slightly off-topic), would you consider a multi-user web UI?
I know that it would (at least in part) conflict with the `SDoc` / `doorstop` approach to store requirements as text files and make use of existing development tool (as git) to manage versions and base-lines. On the other hand, `SDoc` could be used as a backbone for a multi-user application (sort of drop-in replacement of commercial requirements management tools, that use a client-server approach).
Especially when dealing with system engineering requirements (rather than SW requirements), the common user really appreciates the per-document approach of `SDoc`, but has little to no experience with version control systems. Extending `SDoc` in that direction could make happy tons of users!
For sure, adopting such approach would add complexity to the development of the web UI, for example to support a per-user workspace and to implement some way to check-in/out (lock) requirements or to handle conflicts.

**Reply by fkromer** [reply]
@stanislaw @gioboske I've already started to develop something for `strictdoc` before I found out about `reqifviewer`. There is not much developed yet. However I'll provide the source code that others can make progress based on it.

**Reply by stanislaw** [reply]
> However I'll provide the source code that others can make progress based on it.

@fkromer please go ahead! I would like to take a look at your code setup so far. Maybe it is not that hard to get to a minimum viable UI solution.

**Reply by stanislaw** [reply]
@fkromer I feel like I could give this idea a try. I would go with a web interface based on Vue.JS connecting to FastAPI connecting to a given Git repository with SDoc files via pygit interface.

I would start with simply browsing, then adding and then editing requirements. The backend could commit the edits to the Git repository via pygit.

Is there a skeleton you could share that I could check regarding the tech stack suggested by you, or should I try my own way?

Thanks.

**Reply by fkromer** [reply]
@stanislaw I‘ll provide what I‘ve done so far on monday.

**Reply by fkromer** [reply]
Feel free to use https://github.com/fkromer/strictdoc/tree/server .

**Reply by stanislaw** [reply]
Thanks for publishing this. I have started doing something very similar, and I think I got quite far already. Learning the Vue.js framework on the go, I will need to spend some time to get up to speed. The web-related work is a separate repository that will stay private for some time until I have something decent to show.

If there is anyone who is reading now or will read this in the future, and you have frontend experience, let me know if you would like to help. 🏗️ 👷

**Comment by stanislaw** [comment]
The web user interface and the web backend are now in place. 

See this comment: https://github.com/strictdoc-project/strictdoc/issues/563#issuecomment-1478416181.

See the initial design documentation: https://strictdoc.readthedocs.io/en/stable/strictdoc_21_design.html.
